### PR TITLE
feature: add verbose flag and tracing library

### PR DESCRIPTION
**Pull request by rslabbert**
As discussed in [discord](https://discord.com/channels/968932220549103686/969291218347524238/1043744808709140530) this adds a verbose flag which integrates with the tracing library.

I'm not 100% happy with how the verbose flag gets enabled and would be happy to discuss any approaches for making that cleaner. Right now some early debug/trace logs might get missed out due to when the flag gets actioned.

See #772 for actual instrumentation.

# Checklist

If applicable:
- [X] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)

### CODE REVIEW

File: src/main.rs
Comment (martinvonz) [inline src/main.rs:25]: There are two other examples of how to customize the `jj` binary in `examples/`. I suppose we'll want `-v` to work for those without duplicating this code, so some restructuring is needed. The call to `jujutsu::cleanup_guard::init()` would also make sense to move into a shared place. I'm not sure how to structure it. Maybe some `CliRunner` struct that could keep the `(filter, reload_log_filter)` pair? Feel free to just leave a TODO about if you prefer to do it after this PR has landed.

Comment (rslabbert) [reply src/main.rs:29]: I've added a TODO. I'll need to dig a bit more into how the cleanup guard and the Ui code works before I can make a reasonable change there I think.

File: tests/test_global_opts.rs
Comment (martinvonz) [inline tests/test_global_opts.rs:238]: Can you add some test specifically for verbose logging? Just making sure that we actually get verbose logging when we ask for it. Maybe also with the EnvFilter?

Comment (rslabbert) [reply tests/test_global_opts.rs:238]: `test_verbose_logging_enabled` added at the end of the `test_global_opts` file. Turning on verbose logging now also logs out a "verbose logging enabled" message in order to properly test this (and it might give users more comfort if they're not seeing any other verbose logs).

File: src/main.rs
Comment (yuja) [inline src/main.rs:55]: Maybe better to use stderr instead?
https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/struct.Layer.html#method.with_writer

Comment (rslabbert) [reply src/main.rs]: Good shout, I've updated it to use stderr, which should make it more reliable in piping and such workflows.

Comment (yuja) [reply src/main.rs]: thanks!

**Review by yuja — APPROVED** [review]

**Review by martinvonz — APPROVED** [review]
Thanks again. Feel free to merge whenever you want.

File: tests/test_global_opts.rs
Comment (Ralith) [inline tests/test_global_opts.rs:243]: I don't see this test running in the CI output. Am I missing something, or is CI not hitting these for some reason? I ask because I'm not sure if the tracing `fmt` layer will include ANSI on windows by default, which might make this test fragile.

**Review by Ralith — CHANGES_REQUESTED** [review]

File: tests/test_global_opts.rs
Comment (rslabbert) [reply tests/test_global_opts.rs:243]: I can find them in the build runs for Windows and Linux (line 600 and line 567 of latest actions respectively). Make sure to use the actions search bar and not browser search.

**Review by Ralith — APPROVED** [review]

**Comment by martinvonz** [comment]
@rslabbert, feel free to merge. We generally let the author decide when to merge in this project. We use GitHub's branch protections to make sure that you don't merge too early. FYI, you need one approval, passing checks, and no unresolved comments.

**Pull request merged**
